src/tools/user-agent-explainer/user-agent-explainer.vue

```diff
@@ -0,0 +1,141 @@
+<template>
+  <div>
+    <n-space justify="center">
+      <n-form-item label="Use user agent from local browser?" label-placement="left" :show-feedback="false">
+        <n-switch v-model:value="useCurrentUserAgent" @click="useCurrentUserAgentClick" />
+      </n-form-item>
+    </n-space>
+    <n-form-item label="User agent string" v-bind="validationAttrs">
+      <n-input v-model:value="ua" :disabled="useCurrentUserAgent" />
+    </n-form-item>
+
+    <user-agent-result-cards :user-agent-info="userAgentInfo" :sections="sections" />
+  </div>
+</template>
+
+<script setup lang="ts">
+import { computed, PropType } from 'vue';
+import { useStorage } from '@vueuse/core';
+import { UAParser } from 'ua-parser-js';
+import { withDefaultOnError } from '@/utils/defaults';
+import { useValidation } from '@/composable/validation';
+import { isNotThrowing } from '@/utils/boolean';
+import { Adjustments, Browser, Cpu, Devices, Engine } from '@vicons/tabler';
+import UserAgentResultCards from './user-agent-result-cards.vue';
+
+const ua = useStorage('user-agent-explainer:ua', navigator.userAgent);
+const useCurrentUserAgent = useStorage('user-agent-explainer.useCurrentUserAgent', true);
+
+const getUserAgentInfo = (userAgent: string) => new UAParser(userAgent.trim());
+
+const userAgentInfo = computed(() => withDefaultOnError(() => getUserAgentInfo(ua.value), undefined));
+
+const { attrs: validationAttrs } = useValidation({
+  source: ua,
+  rules: [
+    {
+      message: 'We cannot parse this user agent string, check the format',
+      validator: (value) => isNotThrowing(() => getUserAgentInfo(value.trim())),
```

Comment:
Of memory, I think that ua-parser-js is really resilient and never throw event if you pass full gargbage
Have you managed to test this validation ?

Reply:
I haven't found a way to make ua-parser-js throw an error. So I removed the validation.